src/amo/components/AddonQRCode/index.js

```diff
@@ -0,0 +1,184 @@
+/* @flow */
+import invariant from 'invariant';
+import * as React from 'react';
+import { compose } from 'redux';
+
+import {
+  CLIENT_APP_ANDROID,
+  DEFAULT_UTM_SOURCE,
+  DEFAULT_UTM_MEDIUM,
+} from 'amo/constants';
+import Link from 'amo/components/Link';
+import Button from 'amo/components/Button';
+import IconXMark from 'amo/components/IconXMark';
+import { replaceStringsWithJSX } from 'amo/i18n/utils';
+import tracking from 'amo/tracking';
+import { getAddonURL } from 'amo/utils';
+import { addQueryParams } from 'amo/utils/url';
+import translate from 'amo/i18n/translate';
+import type { AddonType } from 'amo/types/addons';
+import type { ElementEvent } from 'amo/types/dom';
+import type { I18nType } from 'amo/types/i18n';
+
+import './styles.scss';
+import qrCode229918 from './img/229918.png';
+import qrCode328839 from './img/328839.png';
+import qrCode506646 from './img/506646.png';
+import qrCode520576 from './img/520576.png';
+import qrCode521554 from './img/521554.png';
+import qrCode607454 from './img/607454.png';
+import qrCode627490 from './img/627490.png';
+import qrCode722 from './img/722.png';
+import qrCode735894 from './img/735894.png';
+import qrCode811592 from './img/811592.png';
+import qrCode824288 from './img/824288.png';
+import qrCode855413 from './img/855413.png';
+import qrCode866226 from './img/866226.png';
+import qrCode869140 from './img/869140.png';
+import qrCode953945 from './img/953945.png';
+import qrCode9609 from './img/9609.png';
+
+export const qrCodeSrcs = {
+  '229918': qrCode229918,
+  '328839': qrCode328839,
+  '506646': qrCode506646,
+  '520576': qrCode520576,
+  '521554': qrCode521554,
+  '607454': qrCode607454,
+  '627490': qrCode627490,
+  '722': qrCode722,
+  '735894': qrCode735894,
+  '811592': qrCode811592,
+  '824288': qrCode824288,
+  '855413': qrCode855413,
+  '866226': qrCode866226,
+  '869140': qrCode869140,
+  '953945': qrCode953945,
+  '9609': qrCode9609,
+};
```

Comment:
I can't help thinking there should be a more efficient way of doing this than importing each image file and then building a data structure referencing the imports but 🤷‍♂️ experiment I guess.

Reply:
Is just a list of the ids and loading the specific image on demand possible/practical?

Reply:
I don't really like all the imports either. I am wondering if we could use URL paths instead (and find a way to copy the files to the `dist/` folder with webpack).

Reply:
I agree it's ugly, but I'm not sure of another way to do it with the files living locally. It's possible we could do something fancy with Webpack, but is it worth it for an experiment? 

I think if the experiment is successful and we decide to move forward with this, we quite likely will end up generating these on the server and returning a URL to the image file in an API response, much like we do for other add-on images. For this reason I'm not convinced it's worth putting a lot of time into a better temporary solution. Having said that, if you have any specific suggestions of how this might work I'm happy to try.

One advantage to this approach, as I mentioned in a separate comment, is that at least we have a definitive list of all add-ons for which we can possibly display a QR code. Using this list to control the display of links limits the possibility of displaying a link to an invalid QR code, which would then either display nothing, or throw an exception. Without this safety net, we'd need extra checks to make sure we're not trying to display an invalid add-on in the panel.

I know it's all a bit hacky, but it actually works quite well, and is temporary.

Reply:
I'll play around a bit with Webpack to see if I can do something with it. If it's not too much work then I think it's worth it. It's just that messing with Webpack scares me. 😬

Reply:
Will and I have been discussing this on Slack. I was able to get Webpack to copy the files to the `/dist` folder, but I'm not sure what it's going to take to make them usable both locally and when deployed. He has offered to take a look. Here's a diff of the changes I made:

```diff
diff --git a/package.json b/package.json
index ca155dbc8..e6ca2b05c 100644
--- a/package.json
+++ b/package.json
@@ -172,6 +172,7 @@
     "compression": "1.7.4",
     "config": "3.3.6",
     "connected-react-router": "5.0.1",
+    "copy-webpack-plugin": "9.0.1",
     "core-js": "3.17.3",
     "deep-eql": "4.0.0",
     "deepcopy": "2.1.0",
diff --git a/webpack-common.js b/webpack-common.js
index eaa645691..4a20e6105 100644
--- a/webpack-common.js
+++ b/webpack-common.js
@@ -1,6 +1,9 @@
 /* eslint-disable import/no-extraneous-dependencies */
+import path from 'path';
+
 import autoprefixer from 'autoprefixer';
 import CircularDependencyPlugin from 'circular-dependency-plugin';
+import CopyPlugin from 'copy-webpack-plugin';
 import config from 'config';
 import MiniCssExtractPlugin from 'mini-css-extract-plugin';
 import webpack from 'webpack';
@@ -143,6 +146,20 @@ export function getPlugins({ withBrowserWindow = true } = {}) {
       /locale$/,
       new RegExp(`^\\.\\/.*?\\/amo\\.js$`),
     ),
+    new CopyPlugin({
+      patterns: [
+        {
+          from: path.resolve(
+            __dirname,
+            'src',
+            'amo',
+            'components',
+            'AddonQRCode',
+            'img',
+          ),
+        },
+      ],
+    }),
   ];

   if (withBrowserWindow) {
```

Reply:
On top of this PR, and thanks to your patch above, I made some changes to simplify the code a bit and remove all the imports. This works in local dev and it should work on hosted envs as well:

```diff
diff --git a/config/default.js b/config/default.js
index c2ccb1a9e..b806ef849 100644
--- a/config/default.js
+++ b/config/default.js
@@ -111,6 +111,7 @@ module.exports = {
     'mozillaUserId',
     'restrictSearchResultsToAppVersion',
     'rtlLangs',
+    'staticPath',
     'trackingEnabled',
     'trackingId',
     'trackingSendInitPageView',
diff --git a/config/development-local.js b/config/development-local.js
index e700f3ed7..a2e9f444d 100644
--- a/config/development-local.js
+++ b/config/development-local.js
@@ -1,3 +1,5 @@
+import { staticPath } from './lib/shared';
+
 // This config should be used with a local addons-server setup.
 module.exports = {
   apiHost: 'http://olympia.test',
@@ -11,6 +13,10 @@ module.exports = {
 
   mozillaUserId: 10968,
   CSP: false,
+
+  // `config/development.js` overrides the static path to use
+  // webpack-dev-server but we don't want that for this environment.
+  staticPath,
   enableNodeStatics: true,
 
   // See: https://github.com/mozilla/addons-frontend/issues/10545
diff --git a/config/development.js b/config/development.js
index 9c8a0c269..1165749c4 100644
--- a/config/development.js
+++ b/config/development.js
@@ -3,7 +3,7 @@ import { addonsServerDevCDN, analyticsHost, apiDevHost } from './lib/shared';
 
 const webpackServerHost = process.env.WEBPACK_SERVER_HOST || '127.0.0.1';
 const webpackServerPort = 3001;
-const webpackHost = `${webpackServerHost}:${webpackServerPort}`;
+const webpackDevServer = `${webpackServerHost}:${webpackServerPort}`;
 
 module.exports = {
   apiHost: 'http://localhost:3000',
@@ -38,7 +38,10 @@ module.exports = {
   serverPort: 3333,
   webpackServerHost,
   webpackServerPort,
-  webpackHost,
+
+  // In local dev, we serve static files using webpack-dev-server.
+  // We need to remove the protocol because of `yarn amo:dev-https`.
+  staticPath: `//${webpackDevServer}/`,
 
   CSP: {
     directives: {
@@ -47,25 +50,25 @@ module.exports = {
         addonsServerDevCDN,
         analyticsHost,
         apiDevHost,
-        webpackHost,
+        webpackDevServer,
         // This is needed for pino-devtools.
         `${webpackServerHost}:3010`,
       ],
       fontSrc: [
-        webpackHost,
+        webpackDevServer,
       ],
       imgSrc: [
         "'self'",
         'data:',
         addonsServerDevCDN,
-        webpackHost,
+        webpackDevServer,
       ],
       scriptSrc: [
         "'self'",
         // webpack injects inline JS
         "'unsafe-inline'",
         addonsServerDevCDN,
-        webpackHost,
+        webpackDevServer,
         `${analyticsHost}/analytics.js`,
       ],
       styleSrc: [
diff --git a/package.json b/package.json
index ca155dbc8..e6ca2b05c 100644
--- a/package.json
+++ b/package.json
@@ -172,6 +172,7 @@
     "compression": "1.7.4",
     "config": "3.3.6",
     "connected-react-router": "5.0.1",
+    "copy-webpack-plugin": "9.0.1",
     "core-js": "3.17.3",
     "deep-eql": "4.0.0",
     "deepcopy": "2.1.0",
diff --git a/src/amo/components/AddonQRCode/index.js b/src/amo/components/AddonQRCode/index.js
index e2af690b4..851194510 100644
--- a/src/amo/components/AddonQRCode/index.js
+++ b/src/amo/components/AddonQRCode/index.js
@@ -2,6 +2,7 @@
 import invariant from 'invariant';
 import * as React from 'react';
 import { compose } from 'redux';
+import config from 'config';
 
 import {
   CLIENT_APP_ANDROID,
@@ -21,41 +22,12 @@ import type { ElementEvent } from 'amo/types/dom';
 import type { I18nType } from 'amo/types/i18n';
 
 import './styles.scss';
-import qrCode229918 from './img/229918.png';
-import qrCode328839 from './img/328839.png';
-import qrCode506646 from './img/506646.png';
-import qrCode520576 from './img/520576.png';
-import qrCode521554 from './img/521554.png';
-import qrCode607454 from './img/607454.png';
-import qrCode627490 from './img/627490.png';
-import qrCode722 from './img/722.png';
-import qrCode735894 from './img/735894.png';
-import qrCode811592 from './img/811592.png';
-import qrCode824288 from './img/824288.png';
-import qrCode855413 from './img/855413.png';
-import qrCode866226 from './img/866226.png';
-import qrCode869140 from './img/869140.png';
-import qrCode953945 from './img/953945.png';
-import qrCode9609 from './img/9609.png';
-
-export const qrCodeSrcs = {
-  '229918': qrCode229918,
-  '328839': qrCode328839,
-  '506646': qrCode506646,
-  '520576': qrCode520576,
-  '521554': qrCode521554,
-  '607454': qrCode607454,
-  '627490': qrCode627490,
-  '722': qrCode722,
-  '735894': qrCode735894,
-  '811592': qrCode811592,
-  '824288': qrCode824288,
-  '855413': qrCode855413,
-  '866226': qrCode866226,
-  '869140': qrCode869140,
-  '953945': qrCode953945,
-  '9609': qrCode9609,
-};
+
+// TODO: move to config
+const EXPECTED_ADDON_IDS = [
+  229918, 328839, 506646, 520576, 521554, 607454, 627490, 722, 735894, 811592,
+  824288, 855413, 866226, 869140, 953945, 9609,
+];
 
 export const ADDON_QRCODE_CAMPAIGN = 'addon-qr-code';
 export const ADDON_QRCODE_CATEGORY = 'Addon QR Code';
@@ -63,7 +35,7 @@ export const ADDON_QRCODE_CLICK_ACTION = 'addon-qr-code-click';
 export const ADDON_QRCODE_IMPRESSION_ACTION = 'addon-qr-code-impression';
 
 type Props = {|
-  addon: AddonType,
+  addon?: AddonType,
   onDismiss: (e: ElementEvent | null) => void,
 |};
 
@@ -104,15 +76,20 @@ export class AddonQRCodeBase extends React.Component<InternalProps> {
   render(): React.Node {
     const { addon, i18n, onDismiss } = this.props;
 
-    const idAsString = String(addon.id);
-    const imgSrc = qrCodeSrcs[idAsString];
-    invariant(imgSrc, `Could not find a QR code for addonId: ${idAsString}`);
+    if (!addon) {
+      return null;
+    }
+
+    invariant(
+      EXPECTED_ADDON_IDS.includes(addon.id),
+      'add-on should have a QR code',
+    );
 
     let addonUrl = addQueryParams(getAddonURL(addon.slug), {
       utm_source: DEFAULT_UTM_SOURCE,
       utm_medium: DEFAULT_UTM_MEDIUM,
       utm_campaign: ADDON_QRCODE_CAMPAIGN,
-      utm_content: String(idAsString),
+      utm_content: String(addon.id),
     });
     addonUrl = `/${CLIENT_APP_ANDROID}${addonUrl}`;
 
@@ -169,7 +146,7 @@ export class AddonQRCodeBase extends React.Component<InternalProps> {
         <img
           alt={altText}
           className="AddonQRCode-img"
-          src={qrCodeSrcs[idAsString]}
+          src={`${config.get('staticPath')}${addon.id}.png?v=1`}
         />
         <div className="AddonQRCode-label">{labelWithLink}</div>
       </div>
diff --git a/webpack-common.js b/webpack-common.js
index eaa645691..4a20e6105 100644
--- a/webpack-common.js
+++ b/webpack-common.js
@@ -1,6 +1,9 @@
 /* eslint-disable import/no-extraneous-dependencies */
+import path from 'path';
+
 import autoprefixer from 'autoprefixer';
 import CircularDependencyPlugin from 'circular-dependency-plugin';
+import CopyPlugin from 'copy-webpack-plugin';
 import config from 'config';
 import MiniCssExtractPlugin from 'mini-css-extract-plugin';
 import webpack from 'webpack';
@@ -143,6 +146,20 @@ export function getPlugins({ withBrowserWindow = true } = {}) {
       /locale$/,
       new RegExp(`^\\.\\/.*?\\/amo\\.js$`),
     ),
+    new CopyPlugin({
+      patterns: [
+        {
+          from: path.resolve(
+            __dirname,
+            'src',
+            'amo',
+            'components',
+            'AddonQRCode',
+            'img',
+          ),
+        },
+      ],
+    }),
   ];
 
   if (withBrowserWindow) {
diff --git a/webpack.dev.config.babel.js b/webpack.dev.config.babel.js
index 52a484820..700489966 100644
--- a/webpack.dev.config.babel.js
+++ b/webpack.dev.config.babel.js
@@ -30,10 +30,9 @@ export const babelOptions = {
     : babelPlugins,
 };
 
-const webpackHost = config.get('webpackServerHost');
-const webpackPort = config.get('webpackServerPort');
-
-const hmr = `webpack-hot-middleware/client?path=//${webpackHost}:${webpackPort}/__webpack_hmr`;
+const hmr = `webpack-hot-middleware/client?path=${config.get(
+  'staticPath',
+)}__webpack_hmr`;
 
 const entryPoints = { [WEBPACK_ENTRYPOINT]: [hmr, 'amo/client'] };
 
@@ -51,7 +50,7 @@ export default {
     filename: '[name]-[contenthash].js',
     chunkFilename: '[name]-[contenthash].js',
     // We need to remove the protocol because of `yarn amo:dev-https`.
-    publicPath: `//${webpackHost}:${webpackPort}/`,
+    publicPath: config.get('staticPath'),
   },
   module: {
     rules: getRules({ babelOptions, bundleStylesWithJs: true }),
diff --git a/yarn.lock b/yarn.lock
index 20dcd4650..edbe67cd5 100644
--- a/yarn.lock
+++ b/yarn.lock
@@ -3660,6 +3660,19 @@ copy-descriptor@^0.1.0:
   resolved "https://registry.yarnpkg.com/copy-descriptor/-/copy-descriptor-0.1.1.tgz#676f6eb3c39997c2ee1ac3a924fd6124748f578d"
   integrity sha1-Z29us8OZl8LuGsOpJP1hJHSPV40=
 
+copy-webpack-plugin@9.0.1:
+  version "9.0.1"
+  resolved "https://registry.yarnpkg.com/copy-webpack-plugin/-/copy-webpack-plugin-9.0.1.tgz#b71d21991599f61a4ee00ba79087b8ba279bbb59"
+  integrity sha512-14gHKKdYIxF84jCEgPgYXCPpldbwpxxLbCmA7LReY7gvbaT555DgeBWBgBZM116tv/fO6RRJrsivBqRyRlukhw==
+  dependencies:
+    fast-glob "^3.2.5"
+    glob-parent "^6.0.0"
+    globby "^11.0.3"
+    normalize-path "^3.0.0"
+    p-limit "^3.1.0"
+    schema-utils "^3.0.0"
+    serialize-javascript "^6.0.0"
+
 core-js-compat@^3.9.0:
   version "3.9.1"
   resolved "https://registry.yarnpkg.com/core-js-compat/-/core-js-compat-3.9.1.tgz#4e572acfe90aff69d76d8c37759d21a5c59bb455"
@@ -5618,6 +5631,13 @@ glob-parent@^5.0.0, glob-parent@^5.1.0:
   dependencies:
     is-glob "^4.0.1"
 
+glob-parent@^6.0.0:
+  version "6.0.1"
+  resolved "https://registry.yarnpkg.com/glob-parent/-/glob-parent-6.0.1.tgz#42054f685eb6a44e7a7d189a96efa40a54971aa7"
+  integrity sha512-kEVjS71mQazDBHKcsq4E9u/vUzaLcw1A8EtUeydawvIWQCJM0qQ08G1H7/XTjFUulla6XQiDOG6MXSaG0HDKog==
+  dependencies:
+    is-glob "^4.0.1"
+
 glob-parent@~5.1.2:
   version "5.1.2"
   resolved "https://registry.yarnpkg.com/glob-parent/-/glob-parent-5.1.2.tgz#869832c58034fe68a4093c17dc15e8340d8401c4"
@@ -11052,7 +11072,7 @@ send@^0.16.1:
     range-parser "~1.2.0"
     statuses "~1.4.0"
 
-serialize-javascript@6.0.0:
+serialize-javascript@6.0.0, serialize-javascript@^6.0.0:
   version "6.0.0"
   resolved "https://registry.yarnpkg.com/serialize-javascript/-/serialize-javascript-6.0.0.tgz#efae5d88f45d7924141da8b5c3a7a7e663fefeb8"
   integrity sha512-Qr3TosvguFt8ePWqsvRfrKyQXIiW+nGbYpy8XK24NQHE83caxWt+mIymTT19DGFbNWNLfEwsrkSmN64lVWB9ag==
```

Reply:
Thanks for this @willdurand! I added the code locally and it does indeed work. I guess we can test it on dev after this and the next PR are approved and merged. I'm going to make a few more changes to this PR and re-submit for review.

Reply:
@willdurand There is a new patch with these changes in it available for review.